src/coprocessor/service.rs

```diff
@@ -0,0 +1,1083 @@
+                }
+            }
+            REQ_TYPE_CHECKSUM => {
+                let mut checksum = ChecksumRequest::new();
+                box_try!(checksum.merge_from(&mut is));
+                let table_scan = checksum.get_scan_on() == ChecksumScanOn::Table;
+                let req_ctx = ReqContext::new(
+                    req.get_context(),
+                    checksum.get_start_ts(),
+                    table_scan,
+                    self.max_handle_duration,
+                );
+                box move |snap| {
+                    ChecksumContext::new(checksum, ranges, snap, req_ctx)
+                        .map(|ctx| ctx.into_boxed())
+                }
+            }
+            tp => return Err(box_err!("unsupported tp {}", tp)),
+        };
+        Ok(builder)
+    }
+
+    /// Creates a `RequestHandlerBuilder`.
+    #[inline]
+    fn new_request_handler_builder(
+        &self,
+        req: &coppb::Request,
+        is_streaming: bool,
+    ) -> RequestHandlerBuilder {
+        self.try_new_request_handler_builder(req, is_streaming)
+            .unwrap_or_else(|e| box move |_| Ok(cop_util::ErrorRequestHandler::new(e).into_boxed()))
+    }
+
+    #[inline]
+    fn get_batch_row_limit(&self, is_streaming: bool) -> usize {
+        if is_streaming {
+            self.stream_batch_row_limit
+        } else {
+            self.batch_row_limit
+        }
+    }
+
+    #[inline]
+    fn async_snapshot(
+        engine: Box<storage::Engine>,
+        ctx: &kvrpcpb::Context,
+    ) -> impl Future<Item = Box<storage::Snapshot + 'static>, Error = Error> {
+        let (callback, future) = util::future::paired_future_callback();
+        let val = engine.async_snapshot(ctx, callback);
+        future::result(val)
+            .and_then(|_| future.map_err(|cancel| storage::engine::Error::Other(box_err!(cancel))))
+            .and_then(|(_ctx, result)| result)
+            // map engine::Error -> coprocessor::Error
+            .map_err(Error::from)
+    }
+
+    fn handle_unary_request_by_custom_handler(
+        &self,
+        context: kvrpcpb::Context,
+        handler_builder: RequestHandlerBuilder,
+    ) -> impl Future<Item = coppb::Response, Error = ()> {
+        let mut tracker = Tracker::new(&context);
+
+        let engine = self.engine.clone();
+        let priority = readpool::Priority::from(context.get_priority());
+
+        let result = self.read_pool.future_execute(priority, move |ctxd| {
+            Service::async_snapshot(engine, &context)
+                .map_err(|e| (e, None))
+                .and_then(move |snapshot| {
+                    let mut handler: Box<RequestHandler> = match handler_builder(snapshot) {
+                        Ok(handler) => handler,
+                        Err(e) => cop_util::ErrorRequestHandler::new(e).into_boxed(),
+                    };
+
+                    {
+                        let ctx = handler.get_context();
```

Comment:
I prefer init `Tracker` with these arguments so that we can simplify some `Option` in it.

Reply:
I would do so if allowed :) Current implementation is really a workaround because some of the argument are unknown at the beginning of the function and it has to be lazy set.

Reply:
Could we init `Tracker` here?

Reply:
It could, but we will need another variable to track the `req_begin` time. What's worse is that, it has to be tracked outside the tracker because we don't have tracker at that time.

Reply:
So I think what we want is:
```rust
fn Tracker::new(start_time: Instant, ctx, ...) -> Self { ... }
```

Reply:
Yes. I mean, if we use this interface, we are letting `start_time` tracked outside the tracker, which breaks the tracker's encapsulation. It seems to be not elegant as well.